File: android-native/app/src/main/java/flexr/social/app/data/local/Entities.kt
package flexr.social.app.data.local

import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import androidx.room.TypeConverter
import flexr.social.app.domain.model.Gender
import flexr.social.app.domain.model.MatchSummary
import flexr.social.app.domain.model.Message
import flexr.social.app.domain.model.Photo
import flexr.social.app.domain.model.PhotoStatus
import flexr.social.app.domain.model.Profile
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.time.Instant

private val converterJson = Json { ignoreUnknownKeys = true }

@Serializable
data class StoredPhoto(
    val id: String,
    val url: String,
    val thumbUrl: String? = null,
    val position: Int = 0,
    val status: String = "approved",
)

class RoomConverters {

    @TypeConverter
    fun photosToJson(photos: List<StoredPhoto>?): String =
        converterJson.encodeToString(photos ?: emptyList())

    @TypeConverter
    fun jsonToPhotos(raw: String?): List<StoredPhoto> =
        if (raw.isNullOrBlank()) emptyList()
        else runCatching { converterJson.decodeFromString<List<StoredPhoto>>(raw) }.getOrDefault(emptyList())

    @TypeConverter
    fun instantToLong(instant: Instant?): Long? = instant?.toEpochMilli()

    @TypeConverter
    fun longToInstant(millis: Long?): Instant? = millis?.let(Instant::ofEpochMilli)
}

/**
 * Lokaler Spiegel der Match-Liste. Dient als Single Source of Truth für die
 * Oberfläche: Matches und Chats sind sofort sichtbar, auch ohne Netz, und
 * werden im Hintergrund aufgefrischt.
 */
@Entity(tableName = "matches")
data class MatchEntity(
    @PrimaryKey val matchId: String,
    val profileId: String,
    val name: String,
    val age: Int,
    val city: String,
    val gender: String,
    val gym: String,
    val bio: String?,
    val isVerified: Boolean,
    val isOnline: Boolean,
    val distanceKm: Int?,
    val photos: List<StoredPhoto>,
    val unreadCount: Int,
    val lastMessageId: String?,
    val lastMessageContent: String?,
    val lastMessageSenderId: String?,
    val lastMessageAt: Instant?,
    val matchedAt: Instant,
)

@Entity(
    tableName = "messages",
    indices = [Index("matchId"), Index("createdAt")],
)
data class MessageEntity(
    @PrimaryKey val id: String,
    val matchId: String,
    val senderId: String,
    val content: String,
    val createdAt: Instant,
    val readAt: Instant?,
    val wasCensored: Boolean,
    /** true, solange die Nachricht nur lokal existiert (optimistisch gesendet). */
    val isPending: Boolean = false,
)

// ---------- Mapping Entity <-> Domain ----------

fun StoredPhoto.toDomain() = Photo(
    id = id,
    url = url,
    thumbUrl = thumbUrl,
    position = position,
    status = PhotoStatus.from(status),
)

fun Photo.toStored() = StoredPhoto(
    id = id,
    url = url,
    thumbUrl = thumbUrl,
    position = position,
    status = status.name.lowercase(),
)

fun MatchEntity.toDomain(): MatchSummary = MatchSummary(
    matchId = matchId,
    profile = Profile(
        id = profileId,
        name = name,
        age = age,
        city = city,
        gender = Gender.from(gender),
        gym = gym,
        bio = bio,
        isOnline = isOnline,
        isVerified = isVerified,
        distanceKm = distanceKm,
        photos = photos.map { it.toDomain() },
    ),
    lastMessage = lastMessageId?.let { id ->
        Message(
            id = id,
            matchId = matchId,
            senderId = lastMessageSenderId.orEmpty(),
            content = lastMessageContent.orEmpty(),
            createdAt = lastMessageAt ?: matchedAt,
            readAt = null,
            wasCensored = false,
        )
    },
    unreadCount = unreadCount,
    isOnline = isOnline,
)

fun MatchSummary.toEntity(matchedAt: Instant = Instant.now()) = MatchEntity(
    matchId = matchId,
    profileId = profile.id,
    name = profile.name,
    age = profile.age,
    city = profile.city,
    gender = profile.gender.apiValue,
    gym = profile.gym,
    bio = profile.bio,
    isVerified = profile.isVerified,
    isOnline = profile.isOnline,
    distanceKm = profile.distanceKm,
    photos = profile.photos.map { it.toStored() },
    unreadCount = unreadCount,
    lastMessageId = lastMessage?.id,
    lastMessageContent = lastMessage?.content,
    lastMessageSenderId = lastMessage?.senderId,
    lastMessageAt = lastMessage?.createdAt,
    matchedAt = lastMessage?.createdAt ?: matchedAt,
)

fun MessageEntity.toDomain() = Message(
    id = id,
    matchId = matchId,
    senderId = senderId,
    content = content,
    createdAt = createdAt,
    readAt = readAt,
    wasCensored = wasCensored,
)

fun Message.toEntity(isPending: Boolean = false) = MessageEntity(
    id = id,
    matchId = matchId,
    senderId = senderId,
    content = content,
    createdAt = createdAt,
    readAt = readAt,
    wasCensored = wasCensored,
    isPending = isPending,
)

File: ios/FLEXR/Data/Local/Entities.swift
import Foundation
import SwiftData

/// Fotoeintrag, wie er am Match hängt. SwiftData speichert ihn als JSON-Blob —
/// eine eigene Tabelle wäre für eine reine Cache-Kopie unnötiger Aufwand.
struct StoredPhoto: Codable, Hashable {
    var id: String
    var url: String
    var thumbURL: String?
    var position: Int
    var status: String

    init(_ photo: Photo) {
        id = photo.id
        url = photo.url
        thumbURL = photo.thumbURL
        position = photo.position
        status = photo.status.rawValue
    }

    func toDomain() -> Photo {
        Photo(id: id, url: url, thumbURL: thumbURL, position: position, status: PhotoStatus(raw: status))
    }
}

/// Lokaler Spiegel der Match-Liste.
///
/// Dient als Single Source of Truth für die Oberfläche: Matches und Chats sind
/// sofort sichtbar, auch ohne Netz, und werden im Hintergrund aufgefrischt.
@Model
final class MatchEntity {

    @Attribute(.unique) var matchID: String
    var profileID: String
    var name: String
    var age: Int
    var city: String
    var gender: String
    var gym: String
    var bio: String?
    var isVerified: Bool
    var isOnline: Bool
    var distanceKm: Int?
    var photosJSON: Data
    var unreadCount: Int
    var lastMessageID: String?
    var lastMessageContent: String?
    var lastMessageSenderID: String?
    var lastMessageAt: Date?
    /// Sortierschlüssel: zuletzt geschriebene Unterhaltung zuerst, Matches ohne
    /// Nachricht nach ihrem Match-Zeitpunkt — wie im Backend.
    var sortedAt: Date
    /// Filter des „Chats"-Tabs. Mit Vorgabewert, damit SwiftData das Feld einer
    /// bereits angelegten Datei leichtgewichtig ergänzen kann statt die Datei
    /// zu verwerfen (siehe FlexrStore).
    var inChats: Bool = false

    init(
        matchID: String,
        profileID: String,
        name: String,
        age: Int,
        city: String,
        gender: String,
        gym: String,
        bio: String?,
        isVerified: Bool,
        isOnline: Bool,
        distanceKm: Int?,
        photosJSON: Data,
        unreadCount: Int,
        lastMessageID: String?,
        lastMessageContent: String?,
        lastMessageSenderID: String?,
        lastMessageAt: Date?,
        sortedAt: Date,
        inChats: Bool
    ) {
        self.matchID = matchID
        self.profileID = profileID
        self.name = name
        self.age = age
        self.city = city
        self.gender = gender
        self.gym = gym
        self.bio = bio
        self.isVerified = isVerified
        self.isOnline = isOnline
        self.distanceKm = distanceKm
        self.photosJSON = photosJSON
        self.unreadCount = unreadCount
        self.lastMessageID = lastMessageID
        self.lastMessageContent = lastMessageContent
        self.lastMessageSenderID = lastMessageSenderID
        self.lastMessageAt = lastMessageAt
        self.sortedAt = sortedAt
        self.inChats = inChats
    }
}

@Model
final class MessageEntity {

    @Attribute(.unique) var messageID: String
    var matchID: String
    var senderID: String
    var content: String
    var createdAt: Date
    var readAt: Date?
    var wasCensored: Bool
    /// true, solange die Nachricht nur lokal existiert (optimistisch gesendet).
    var isPending: Bool

    init(
        messageID: String,
        matchID: String,
        senderID: String,
        content: String,
        createdAt: Date,
        readAt: Date?,
        wasCensored: Bool,
        isPending: Bool = false
    ) {
        self.messageID = messageID
        self.matchID = matchID
        self.senderID = senderID
        self.content = content
        self.createdAt = createdAt
        self.readAt = readAt
        self.wasCensored = wasCensored
        self.isPending = isPending
    }
}

// MARK: - Mapping Entity <-> Domäne

extension MatchEntity {

    var photos: [StoredPhoto] {
        (try? JSONDecoder().decode([StoredPhoto].self, from: photosJSON)) ?? []
    }

    func toDomain() -> MatchSummary {
        MatchSummary(
            matchID: matchID,
            profile: Profile(
                id: profileID,
                name: name,
                age: age,
                city: city,
                gender: Gender(raw: gender),
                gym: gym,
                bio: bio,
                isOnline: isOnline,
                isVerified: isVerified,
                distanceKm: distanceKm,
                photos: photos.map { $0.toDomain() }
            ),
            lastMessage: lastMessageID.map { id in
                Message(
                    id: id,
                    matchID: matchID,
                    senderID: lastMessageSenderID ?? "",
                    content: lastMessageContent ?? "",
                    createdAt: lastMessageAt ?? sortedAt,
                    readAt: nil,
                    wasCensored: false
                )
            },
            unreadCount: unreadCount,
            isOnline: isOnline,
            inChats: inChats
        )
    }

    /// Serverstand in einen bestehenden Datensatz übernehmen.
    func apply(_ summary: MatchSummary, matchedAt: Date = Date()) {
        profileID = summary.profile.id
        name = summary.profile.name
        age = summary.profile.age
        city = summary.profile.city
        gender = summary.profile.gender.apiValue
        gym = summary.profile.gym
        bio = summary.profile.bio
        isVerified = summary.profile.isVerified
        isOnline = summary.isOnline
        distanceKm = summary.profile.distanceKm
        photosJSON = MatchEntity.encode(summary.profile.photos)
        unreadCount = summary.unreadCount
        lastMessageID = summary.lastMessage?.id
        lastMessageContent = summary.lastMessage?.content
        lastMessageSenderID = summary.lastMessage?.senderID
        lastMessageAt = summary.lastMessage?.createdAt
        sortedAt = summary.lastMessage?.createdAt ?? matchedAt
        inChats = summary.inChats
    }

    static func make(_ summary: MatchSummary, matchedAt: Date = Date()) -> MatchEntity {
        MatchEntity(
            matchID: summary.matchID,
            profileID: summary.profile.id,
            name: summary.profile.name,
            age: summary.profile.age,
            city: summary.profile.city,
            gender: summary.profile.gender.apiValue,
            gym: summary.profile.gym,
            bio: summary.profile.bio,
            isVerified: summary.profile.isVerified,
            isOnline: summary.isOnline,
            distanceKm: summary.profile.distanceKm,
            photosJSON: encode(summary.profile.photos),
            unreadCount: summary.unreadCount,
            lastMessageID: summary.lastMessage?.id,
            lastMessageContent: summary.lastMessage?.content,
            lastMessageSenderID: summary.lastMessage?.senderID,
            lastMessageAt: summary.lastMessage?.createdAt,
            sortedAt: summary.lastMessage?.createdAt ?? matchedAt,
            inChats: summary.inChats
        )
    }

    static func encode(_ photos: [Photo]) -> Data {
        (try? JSONEncoder().encode(photos.map(StoredPhoto.init))) ?? Data("[]".utf8)
    }
}

extension MessageEntity {

    func toDomain() -> Message {
        Message(
            id: messageID,
            matchID: matchID,
            senderID: senderID,
            content: content,
            createdAt: createdAt,
            readAt: readAt,
            wasCensored: wasCensored
        )
    }

    static func make(_ message: Message, isPending: Bool = false) -> MessageEntity {
        MessageEntity(
            messageID: message.id,
            matchID: message.matchID,
            senderID: message.senderID,
            content: message.content,
            createdAt: message.createdAt,
            readAt: message.readAt,
            wasCensored: message.wasCensored,
            isPending: isPending
        )
    }
}
